File: Source/app/src/main/java/com/amgregoire/mangafeed/MangaFeed.java
package com.amgregoire.mangafeed;

import android.app.Application;
import android.support.design.widget.Snackbar;
import android.view.View;
import android.widget.Toast;

import com.amgregoire.mangafeed.Common.MangaEnums;
import com.amgregoire.mangafeed.Common.WebSources.Base.SourceBase;
import com.amgregoire.mangafeed.Common.WebSources.FunManga;
import com.amgregoire.mangafeed.Common.WebSources.MangaEden;
import com.amgregoire.mangafeed.Common.WebSources.MangaHere;
import com.amgregoire.mangafeed.Common.WebSources.ReadLight;
import com.amgregoire.mangafeed.Models.Chapter;
import com.amgregoire.mangafeed.Utils.MangaDB;
import com.amgregoire.mangafeed.Utils.MangaLogger;
import com.amgregoire.mangafeed.Utils.RxBus;
import com.amgregoire.mangafeed.Utils.SharedPrefs;
import com.bumptech.glide.request.target.ViewTarget;
import com.squareup.picasso.Cache;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.List;
import java.util.concurrent.Executors;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by Andy Gregoire on 3/8/2018.
 */

public class MangaFeed extends Application
{
    public final static String TAG = MangaFeed.class.getSimpleName();

    private static MangaFeed mInstance;
    private RxBus mBus;

    private ArrayList<Chapter> mCurrentChapters;
    private Picasso mPicasso;

    public MangaFeed()
    {
        mInstance = this;
    }

    @Override
    public void onCreate()
    {
        super.onCreate();
        ViewTarget.setTagId(R.id.glide_tag);
        mBus = new RxBus();

        mPicasso = new Picasso.Builder(this).executor(Executors.newSingleThreadExecutor()).memoryCache(Cache.NONE).indicatorsEnabled(true).build();
        mPicasso.setIndicatorsEnabled(true);
        Picasso.setSingletonInstance(mPicasso);

        MangaDB.getInstance().createDB(); // Copy pre-loaded database if not already done.
        MangaDB.getInstance().initDao();
        updateCatalogs(false); // check if we should update local database on application open

        if(BuildConfig.VERSION_CODE < 2)
        {
            updateCatalogs(true);
        }
    }

    /***
     * This function returns the MangaFeed application instance
     *
     * @return
     */
    public static synchronized MangaFeed getInstance()
    {
        return mInstance;
    }

    /***
     * This function returns the application RxBus instance.
     *
     * @return
     */
    public RxBus rxBus()
    {
        return mBus;
    }

    /***
     * This function creates and shows a SHORT toast message.
     *
     * @param message
     */
    public void makeToastShort(String message)
    {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
    }

    /***
     * This function creates and shows a LONG toast message.
     *
     * @param message
     */
    public void makeToastLong(String message)
    {
        Toast.makeText(this, message, Toast.LENGTH_LONG).show();
    }

    /***
     * This function creates and shows a short snackbar message.
     *
     * TODO: might remove and just use toasts.
     * @param v
     * @param message
     */
    public void makeSnackBarShort(View v, String message)
    {
        Snackbar.make(v, message, Snackbar.LENGTH_LONG).show();
    }

    /***
     * This function returns the current sources source type
     *
     * NOVEL, MANGA
     *
     * @return
     */
    public MangaEnums.SourceType getCurrentSourceType()
    {
        return MangaEnums.Source.valueOf(SharedPrefs.getSavedSource()).getSource().getSourceType();
    }

    /***
     * This function returns the current source.
     *
     * MangaEden, MangaHere. FunManga, ReadLight
     *
     * @return
     */
    public SourceBase getCurrentSource()
    {
        return MangaEnums.Source.valueOf(SharedPrefs.getSavedSource()).getSource();
    }

    /***
     * This function returns a new source specified by the source classes TAG.
     *
     * @param tag
     * @return
     */
    public SourceBase getSourceByTag(String tag)
    {
        if (tag.equals(FunManga.TAG))
        {
            return new FunManga();
        }
        else if (tag.equals(MangaEden.TAG))
        {
            return new MangaEden();
        }
        else if (tag.equals(MangaHere.TAG))
        {
            return new MangaHere();
        }
        else
        {
            return new ReadLight();
        }
    }

    /***
     * This function returns a new source specified by the source classes TAG.
     *
     * @param url
     * @return
     */
    public SourceBase getSourceByUrl(String url)
    {
        if (url.contains(FunManga.URL))
        {
            return new FunManga();
        }
        else if (url.contains(MangaEden.URL))
        {
            return new MangaEden();
        }
        else if (url.contains(MangaHere.URL))
        {
            return new MangaHere();
        }
        else
        {
            return new ReadLight();
        }
    }

    public void setCurrentChapters(List<Chapter> chapters)
    {
        mCurrentChapters = new ArrayList<>(chapters);
        Collections.reverse(mCurrentChapters);
    }

    public List<Chapter> getCurrentChapters()
    {
        return mCurrentChapters;
    }

    /***
     * This function updates source catalogs items on the local database, adding any missing items.
     * Currently it is set to update no more than once a week.
     *
     */
    public void updateCatalogs(boolean isForceUpdate)
    {
        int lWeekSeconds = 604800;
        int lWeekMs = lWeekSeconds * 1000;

        // Check if we updated in the last week, if we have we'll skip.
        Date lLowerLimit = new Date(SharedPrefs.getLastCatalogUpdate().getTime() + lWeekMs);
        if (lLowerLimit.before(new Date()) || isForceUpdate)
        {
            Observable.create((ObservableEmitter<SourceBase> subscriber) ->
            {
                try
                {
                    SharedPrefs.setLastCatalogUpdate();
                    MangaEnums.Source[] lSources = MangaEnums.Source.values();

                    for (MangaEnums.Source source : lSources)
                    {
                        source.getSource().updateLocalCatalog();
                        subscriber.onNext(source.getSource());
                    }
                    subscriber.onComplete();
                }
                catch (Exception ex)
                {
                    subscriber.onError(ex);
                }
            }).subscribeOn(Schedulers.computation()).subscribe(
                    source -> source.updateLocalCatalog(), // onNext
                    throwable -> MangaLogger.logError(TAG, throwable.getMessage()) // onError
            );
        }
    }
}

File: Source/app/src/main/java/com/amgregoire/mangafeed/MangaFeed.kt
package com.amgregoire.mangafeed

import android.app.Application
import android.widget.Toast
import com.amgregoire.mangafeed.Common.MangaEnums
import com.amgregoire.mangafeed.Common.WebSources.Base.SourceBase
import com.amgregoire.mangafeed.Common.WebSources.FunManga
import com.amgregoire.mangafeed.Common.WebSources.MangaEden
import com.amgregoire.mangafeed.Common.WebSources.MangaHere
import com.amgregoire.mangafeed.Common.WebSources.ReadLight
import com.amgregoire.mangafeed.Models.DbChapter
import com.amgregoire.mangafeed.Utils.MangaDB
import com.amgregoire.mangafeed.Utils.RxBus
import com.amgregoire.mangafeed.Utils.SharedPrefs
import com.amgregoire.mangafeed.v2.FunMangaCookiePreferences
import com.amgregoire.mangafeed.v2.ReadLightCookiePreferences
import com.amgregoire.mangafeed.v2.UserPreferences
import com.amgregoire.mangafeed.v2.di.component.AppComponent
import com.amgregoire.mangafeed.v2.di.component.DaggerAppComponent
import com.amgregoire.mangafeed.v2.di.module.ApiModule
import com.amgregoire.mangafeed.v2.di.module.ApplicationModule
import com.amgregoire.mangafeed.v2.di.module.RoomModule
import com.amgregoire.mangafeed.v2.extension.fromJson
import com.amgregoire.mangafeed.v2.extension.toJson
import com.amgregoire.mangafeed.v2.model.domain.User
import com.amgregoire.mangafeed.v2.model.domain.UserLibrary
import com.amgregoire.mangafeed.v2.repository.local.LocalMangaRepository
import com.amgregoire.mangafeed.v2.service.Logger
import com.amgregoire.mangafeed.v2.ui.LoginActivity
import com.amgregoire.mangafeed.v2.usecase.GetUserUseCase
import com.amgregoire.mangafeed.v2.usecase.SignOutUseCase
import com.bumptech.glide.request.target.ViewTarget
import io.reactivex.disposables.CompositeDisposable
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.util.*
import javax.inject.Inject


/**
 * Created by Andy Gregoire on 3/8/2018.
 */
val ioScope = CoroutineScope(Dispatchers.IO)
val uiScope = CoroutineScope(Dispatchers.Main)

val currentSource: SourceBase
    get() = MangaEnums.Source.valueOf(SharedPrefs.getSavedSource()).source

val appComponent: AppComponent = MangaFeed.app.appComponent

class MangaFeed : Application()
{
    private var compositeDisposable = CompositeDisposable()
    private val userPreferences by lazy { UserPreferences(app) }

    @Deprecated("v1")
    private var mCurrentDbChapters: ArrayList<DbChapter>? = null

    var user: User?
        get()
        {
            return userPreferences.user?.fromJson()
        }
        set(value)
        {
            if (value != null)
            {
                userPreferences.user = value.toJson()
                isSignedIn = true
            }
            else userPreferences.clear()
        }

    var isSignedIn: Boolean
        get()
        {
            return userPreferences.isSignedIn
        }
        set(value)
        {
            userPreferences.isSignedIn = value
        }

    var userLibrary: UserLibrary? = null
        set(value)
        {
            ioScope.launch {
                val localMangaRepository = LocalMangaRepository()
                value?.library?.forEach {
                    localMangaRepository.getManga(it.link, it.source)?.apply { this.id = it.id }
                            ?.let {
                                localMangaRepository.putManga(it)
                            }
                }
            }
            field = value
        }

    fun cookiePreferences() = when (currentSource.sourceName)
    {
        ReadLight.SourceKey -> ReadLightCookiePreferences(app)
        else -> FunMangaCookiePreferences(app)
    }


    /************************************************************************
     *
     * Dagger
     *
     ***********************************************************************/
    val appComponent: AppComponent by lazy {
        DaggerAppComponent.builder()
                .applicationModule(ApplicationModule(this))
                .roomModule(RoomModule(this))
                .apiModule(ApiModule())
                .build()
    }

    @Inject lateinit var db: MangaDB
    @Inject lateinit var prefs: SharedPrefs

    /************************************************************************
     *
     *
     *
     ***********************************************************************/

    val currentSourceType: MangaEnums.SourceType
        get() = MangaEnums.Source.valueOf(SharedPrefs.getSavedSource()).source.sourceType

    val currentSource: SourceBase
        get() = MangaEnums.Source.valueOf(SharedPrefs.getSavedSource()).source

    @Deprecated("v1")
    var currentDbChapters: List<DbChapter>?
        get() = mCurrentDbChapters
        set(chapters)
        {
            mCurrentDbChapters = ArrayList(chapters)
            mCurrentDbChapters!!.reverse()
        }

    @Deprecated("v1")
    private var mBus: RxBus? = null

    @Deprecated("v1")
    fun rxBus(): RxBus?
    {
        return mBus
    }

    /************************************************************************
     *
     * Dagger
     *
     ***********************************************************************/

    override fun onCreate()
    {
        super.onCreate()
        mangaApp = this
        cookiePreferences().clear()

        appComponent.inject(this)

        ViewTarget.setTagId(R.id.glide_tag)

        mBus = RxBus()

        updateUser()
    }

    fun logout()
    {
        user?.let {
            SignOutUseCase().signOut(
                    accessToken = it.accessToken,
                    result = { Logger.debug("Logout success") },
                    error = { Logger.debug("Logout failed") }
            )
        }

        this.user = null
        this.userLibrary = null
        startActivity(LoginActivity.newInstance(this))
    }

    /***
     * This function creates and shows a SHORT toast message.
     *
     * @param message
     */
    fun makeToastShort(message: String)
    {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
    }

    /***
     * This function creates and shows a LONG toast message.
     *
     * @param message
     */
    fun makeToastLong(message: String)
    {
        Toast.makeText(this, message, Toast.LENGTH_LONG).show()
    }

    /***
     * This function returns a new source specified by the source classes TAG.
     *
     * @param tag
     * @return
     */
    fun getSourceByTag(tag: String): SourceBase
    {
        return when (tag)
        {
            FunManga.TAG -> FunManga()
            MangaEden.TAG -> MangaEden()
            MangaHere.TAG -> MangaHere()
            else -> ReadLight()
        }
    }

    /***
     * This function returns a new source specified by the source classes TAG.
     *
     * @param url
     * @return
     */
    fun getSourceByUrl(url: String): SourceBase
    {
        return when
        {
            url.contains(FunManga.URL) -> FunManga()
            url.contains(MangaEden.URL) -> MangaEden()
            url.contains(MangaHere.URL) -> MangaHere()
            else -> ReadLight()
        }
    }

    private fun updateUser()
    {
        user ?: return
        GetUserUseCase().user { result ->
            result ?: return@user // TODO :: Determine what to do if failed to get user
            user = result.user
            // TODO :: Use case to sync local database?
            // Maybe have a pop up if there are inconsistencies ?
        }
    }

    companion object
    {
        private var TAG: String = MangaFeed::class.java.simpleName
        private var mangaApp: MangaFeed? = null
        val app get() = checkNotNull(mangaApp)
    }
}
